TBS User-Scoped account management guide (#3336)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/configure-user-scoped-account-management/main/index.md

---

> title: Configure User-Scoped account management

Author: ```suggestion
title: Configure user-scoped account management
```

meta:
  - name: description
    content: The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need a bearer token to update their email and phone authenticators.
layout: Guides
sections:
- main
---

> This guide explains how to set up end user-driven account management in your org using the MyAccount API.

Reviewer: ```suggestion
This guide explains how to set up end-user driven account management in your org using the MyAccount API.
```

<!-- > **Note:** This guide covers a narrow set of use cases that call the MyAccount API. Most use cases are covered by the SDKs. See -->

---
**Learning outcomes**

* Set up your org with the required scopes to call the MyAccount API.

> * Allow your end users to call the API to update their profile and enroll another authenticator after first sign-in.

Reviewer: ```suggestion
* Allow your end users to call the API to update their profile and enroll another authenticator after they first sign in.
```

**What you need**

* [Okta Developer Edition organization](https://developer.okta.com/signup)
* Access to the latest version of the MyAccount API: `/idp/myaccount`
* An existing OpenID Connect client app in Okta for testing in Okta

> [Postman client](https://www.getpostman.com/downloads/) to test requests. See [Get Started with the Okta APIs](https://developer.okta.com/code/rest/) for information on setting up Postman.

Reviewer: ```suggestion
* [Postman client](https://www.getpostman.com/downloads/) to test requests. See [Get Started with the Okta APIs](https://developer.okta.com/code/rest/) for information on setting up Postman.
```

* The MyAccount API Postman collection that allows you to test the API calls that are described in this guide. Click **Run in Postman** to add the collection to your Postman workspace.

> [![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/9cb68745dbf85ae3a871)

Reviewer 2: where did this collection come from (where is it's source)?

Reviewer 3: its from @ericalong-okta 's ticket https://oktainc.atlassian.net/browse/OKTA-495857

Author: Resolving as per our slack discussion.

> > **Note:** To run this Postman collection, you need an end-user access token. Use an SDK to get the token. See [Embedded Auth with SDKs React](https://github.com/okta/okta-auth-js/tree/master/test/apps/react-oie) for a sample app that provides an access token.

Author: ```suggestion
> **Note:** To run this Postman collection, you need an end-user access token. Use an SDK to get the token. See [Embedded Auth with SDKs React](https://github.com/okta/okta-auth-js/tree/master/test/apps/react-oie) for a sample app that provides an access token.
```

> > **Note:** To run this Postman collection, you need an end-user access token. Use a password grant type or SDK to get the token. See [Embedded Auth with SDKs React](https://github.com/okta/okta-auth-js/tree/master/test/apps/react-oie) for a sample app that provides an access token.

Reviewer: Use a password grant type? You aren't implying that they use the Resource Owner Password grant type, right? If so, it's just to walk through this use case, right? Because we discourage the use of the ROPassword grant type. Unless I'm not understanding what you are telling them to use....which is very possible. haha

> > **Note:** To run this Postman collection, you need an end-user access token. Use a password grant type or SDK to get the token. See [Embedded Auth with SDKs React](https://github.com/okta/okta-auth-js/tree/master/test/apps/react-oie) for a sample app that provides an access token.

Author: ```suggestion
> **Note:** To run this Postman collection, you need an end-user access token. Use an SDK to get the token. See [Embedded Auth with SDKs React](https://github.com/okta/okta-auth-js/tree/master/test/apps/react-oie) for a sample app that provides an access token.
```

---

> ## About User-Scoped account management

Reviewer: ```suggestion
## About user-scoped account management
```

> The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need a bearer token to update their profile, or email and phone authenticators. In addition, app developers can call the MyAccount API for active users outside of the authentication context. For example, once a user enrolls in the mandatory email factor and completes authentication, app developers can call the API to enroll the active user with the optional phone authenticator.

Reviewer: ```suggestion
The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need a bearer token to update their profile, or email and phone authenticators. In addition, app developers can call the MyAccount API for active users outside of the authentication context. For example, after a user enrolls in the mandatory email factor and completes authentication, app developers can call the API to enroll the active user with the optional phone authenticator.
```

> The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need an active user session to update their email and phone authenticators. In addition, app developers can call the MyAccount API for active users outside of the authentication context. For example, once a user enrolls in the mandatory email factor and completes authentication, app developers can call the API to enroll the active user with the optional phone authenticator.

Author: ```suggestion
The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need a bearer token to update their email and phone authenticators. In addition, app developers can call the MyAccount API for active users outside of the authentication context. For example, once a user enrolls in the mandatory email factor and completes authentication, app developers can call the API to enroll the active user with the optional phone authenticator.
```

> The MyAccount API now provides user-scoped endpoints that don’t require admin tokens. End users only need an active user session to update their email and phone authenticators. In addition, app developers can call the MyAccount API for active users outside of the authentication context. For example, once a user enrolls in the mandatory email factor and completes authentication, app developers can call the API to enroll the active user with the optional phone authenticator.

Author: Thanks @JayHe-okta I'll change it.

## About the Postman collection

It is up to you how you set up users to call the MyAccount API to manage their account. In this guide, we provide sample API calls using a Postman collection to demonstrate them in a language/platform neutral way.

> > **Note:** To run the Postman collection, you need an end-user access token. Use a password grant type or [SDK](/docs/guides/auth-js/-/main/) to get the token.

Author: ```suggestion
> **Note:** To run the Postman collection, you need an end-user access token. Use an [SDK](/docs/guides/auth-js/-/main/) to get the token.
```

> > **Note:** To run the Postman collection, you need an end-user access token. Use a password grant type or [SDK](/docs/guides/auth-js/-/main/) to get the token.

Reviewer: Use a password grant type? You aren't implying that they use the Resource Owner Password grant type, right? If so, it's just to walk through this use case, right? Because we discourage the use of the ROPassword grant type. Unless I'm not understanding what you are telling them to use....which is very possible. haha

> ## API Versioning

Author: ```suggestion
## API versioning
```

> A valid API version in the `Accept` header is required to access the API. Current version V1.0.0

Reviewer: ```suggestion
A valid API version in the `Accept` header is required to access the API. Current version: V1.0.0
```

```json
Accept: application/json; okta-version=1.0.0
```

## Grant the required scopes

> **Note:** Before setting up your org to send requests to the MyAccount API, make sure that you have an OpenID Connect client app set up in Okta. See 

When you are ready to grant the required scopes, follow these steps:

> 1. Sign in to your Okta organization with your administrator account and go to **Applications** > **Applications**.

Reviewer 3: we should also mention myaccount api also supports custome AS not just orgAS's application scopes. the customAS has all the following scopes enabled.

Author: @JayHe-okta as discussed, I added a note mentioning customAS and these scopes.

1. Open your OpenID Connect client app.

> 1. In the **Okta API Scopes** tab, grant the following scopes:

Reviewer: ```suggestion
1. on the **Okta API Scopes** tab, grant the following scopes:
```

> 1. on the **Okta API Scopes** tab, grant the following scopes:

Author: ```suggestion
1. On the **Okta API Scopes** tab, grant the following scopes:
```

   * For access to both GET and POST/DELETE endpoints:
      * `okta.myAccount.email.manage`
      * `okta.myAccount.phone.manage`
      * `okta.myAccount.profile.manage`
   * For access to GET endpoints only:
      * `okta.myAccount.email.read`
      * `okta.myAccount.phone.read`
      * `okta.myAccount.profile.read`

## Add a secondary email

<ApiOperation method="post" url="/idp/myaccount/emails" />

In this use case, the end user adds a `SECONDARY` email address to their account. The new email address is in UNVERIFIED status.

### Required scope and role

An Okta scope of `okta.myAccount.email.manage` is required to use this endpoint.

> > **Note:** Admin users aren't allowed to call the `/idp/myaccount/emails` POST endpoint.

Reviewer: ```suggestion
> **Note:** Admin users aren't allowed to make a POST request to the `/idp/myaccount/emails` endpoint.
```

### Request path parameters

N/A

### Request query parameters

N/A

### Request body

This API requires a [My Email Request object](#my-email-request-object) as its request body.

```bash
{
  "profile": {

>       "email" : "some.primary.email@okta.com"

Reviewer 2: should this be `secondary`

Author: Yep. Fixed.

>       "email" : "some.primary.email@okta.com"

Author: ```suggestion
      "email" : "some.secondary.email@okta.com"
```

  },
   "role": "SECONDARY",
   "sendEmail": true
}
```

#### Response body

The requested [My Email object](#my-email-object)

#### Error Responses

If an invalid email is passed to `profile` in the request body, the response returns a 400 BAD REQUEST with error code E0000001.

> If the email operation isn't enabled for the request `role` on the org, the response returns a 403 FORBIDDEN with error code E0000038.

Reviewer: ```suggestion
If the email operation isn't enabled for the request `role` in the org, the response returns a 403 FORBIDDEN with error code E0000038.
```

If the email already exists for the current user, the response returns a 409 CONFLICT with error code E0000157.

> ##### Request

Reviewer 2: Is this a new section or it belongs to `## Add a secondary email` ? do we need a new section header?

Author: Same section as "secondary email". I need to adjust the headings.

```bash
curl -v -X POST \
-H "Authorization: Bearer ${token}" \
"https://${yourOktaDomain}/idp/myaccount/emails"
```

#### Response

```json
{
    "id": "5a8de6071e1b94e0f4ec664b9e4869e8",
    "status": "UNVERIFIED",
    "profile": {
        "email": "secondary.email@okta.com"
    },
    "roles": [
        "SECONDARY"
    ],
    "_links": {
        "self": {
            "href": "https://${yourOktaDomain}/idp/myaccount/emails/5a8de6071e1b94e0f4ec664b9e4869e8",
            "hints": {
                "allow": [
                    "GET"
                ]
            }
        },
        "challenge": {
            "href": "https://${yourOktaDomain}/idp/myaccount/emails/5a8de6071e1b94e0f4ec664b9e4869e8/challenge",
            "hints": {
                "allow": [
                    "POST"
                ]
            }
        }
    }
}
```

## Add a phone

<ApiOperation method="post" url="/idp/myaccount/phones" />

> Create an UNVERIFIED status phone for either `SMS` or `CALL` method to the user's MyAccount setting.

Reviewer: ```suggestion
Creates an UNVERIFIED status phone for either the `SMS` or `CALL` method to the user's MyAccount setting.
```

### Required scope and role

An Okta scope of `okta.myAccount.phone.manage` is required to use this endpoint.

> > **Note:** Admin users aren't allowed to call the `/idp/myaccount/phones` POST endpoint.

Reviewer: ```suggestion
> **Note:** Admin users aren't allowed to make a POST request to the `/idp/myaccount/phones` endpoint.
```

#### Request path parameters

N/A

#### Request body

This API requires a [My Phone Request object](#my-phone-request-object) as its request body.

```bash
{
  "profile": {
      "phoneNumber" : "+15555555555"
  },
   "sendCode": true,
   "method": "SMS"
}
```

### Response body

The requested [My Phone object](#my-phone-object)

### Error Responses

If an invalid phone number is passed to `profile` in the request body, the response returns a 400 BAD REQUEST with error code E0000001.

If an invalid `method` is passed in the request body, the response returns a 400 BAD REQUEST with error code E0000001.

> If the phone authenticator isn't enabled for `method` on the org, the response returns a 403 FORBIDDEN with error code E0000038.

Reviewer: ```suggestion
If the phone authenticator isn't enabled for `method` in the org, the response returns a 403 FORBIDDEN with error code E0000038.
```

> If the number of phone factors for the current user already reaches the maximum allowed per user or the phone factor is failed to create, the response returns a 500 INTERNAL SERVER ERROR with error code E0000009.

Reviewer 2: this shouldn't be the case, we should never return a 500. If this is the case on the API then it is a bug and we should fix it.

Author: @zeusandrade-okta that bit of content is from the reference doc.

Author: Fixed, as per our Slack discussion.

If the phone number already exists for the current user, the response returns a 409 CONFLICT with error code E0000157.

> If the call providers fails to send challenge when `sendCode` is true,  the response returns 500 with error code E0000138.

Reviewer: ```suggestion
If the call providers fail to send a challenge when `sendCode` is `true`,  the response returns a 500 with error code E0000138.
```
